app/controllers/image_controller.rb

```diff
@@ -268,7 +230,74 @@ def prev_image
     redirect_to_next_object(:prev, Image, params[:id].to_s)
   end
 
+  private
+
+  def image_exists
+    @image = find_or_goto_index(Image, params[:id].to_s)
```

Comment:
I would prefer to somehow expose the fact that @image is set as a side-effect.  I might add an exclamation, too, to emphasize that it's doing error handling and that the caller should abort if it fails.

Reply:
Are you suggesting that the caller should assign like `@image = image_if_exists!` and then 
```ruby
def image_if_exists!
  find_or_goto_index(Image, params[:id].to_s)
end
```

Reply:
Yes pretty much.  I might personally have preferred "return unless (@image = find_image!)".

This is all just personal aesthetics, you realize that right?  I'm not trying to prove in any absolute, cosmic sense that your solution here is "wrong".  I'm just giving feedback.  These are things I would prefer, and why.

In these sorts of refactors what I like to do for what very little it is worth -- you may approach it differently, so please feel free to ignore! -- is I like to rewrite the main method as a sort of executive bullet-point kind of thing.  It does A, B, C, and D, then returns.  And I like to manage as many globals (instance variables in this case) as possible in the top level.  It feels really good if I can somehow structure it so that each bullet-point is very simple and stand-alone, e.g., relying on 0-2 parameters only.  If it requires more parameters then maybe there's a way to split it?  Sometimes it's not possible, of course.

But then, when you read the top-level method you know exactly what it does, what all of the side-effects are, ideally even _how_ the subroutines are doing the work.  Then I can go and write all the little helper methods (the "bullet-points").

Obviously it was all a great deal easier when we could just throw it all in one big method, no need to worry about passing parameters around or anything, super free-form and forgiving.  But given that we're now being encouraged to write smaller and more modular and all that, I find that this top-down approach can really yield some nice code.  The very rare cases where I've actually done it, that is. :)

Another approach which we can call "bottom-up" is take the big long old method and just slice it up, changing the minimum amount, ideally cutting it into as meaningful chunks as possible.  Certainly _way_ easier and quicker and _way_ less likely to break code.  Like what I did when I "refactored" the observations_controller -- I just literally cut it up into several files but changed nothing fundamentally.  It was meant only to be a baby step in the right direction, and as such it was an improvement and just fine.

I'm not recommending a complete overhaul here.  I just want to make sure when you cut up the existing code the pieces get good names so that we know more or less what the pieces do without having to look up the code.  What's the point of cutting it up into pieces if you still have to read it all to make sense of any of it?  It's just a perverse game that helps nothing except fooling rubocop into thinking it passes all its silly metrics! :)  I understand this is just meant to be a step in the right direction.  And this is.

It's all good.  I'd just like to improve the names a bit, and I think it's good to go.

Reply:
I've re-read this a couple times. You're a good writer and the coding style you describe is exactly what i've been going for.  This clarifies a lot.

The hitch is my bottom-up understanding of the code. I often wade in with the vaguest possible idea what any of the methods do, including the obvious ones. I sometimes bump into blocks of code and discover entire scopes of action on MO that i didn't suspect existed. So i'm usually slogging through, just chopping code into manageable lengths until i have an inkling what they do.

I would write bullet point, directive sub-methods if i had a clear idea of where i was going. I'm grasping for that top-down understanding... and almost never arriving until i've gotten through fixing all the tests, at which point i don't feel like messing with the controller methods again. Hah!

Reply:
Excellent, sounds like we are totally on the same page.  I totally support what we're calling the bottom-up approach here.  There are many times and places where it makes sense.  And this is one of them.

Chop away and worry about making it "make sense" later when it's had time to percolate and you've gotten your head around the whole problem.  I withdraw my (already weak) criticisms.

Reply:
Almost all these suggestions are now committed locally. But i'm getting an error on the thicket of update methods without `xargs`. Question coming...